### ISSUE TITLE
Differences between `emsdk install latest` and `emsdk install sdk-main-64bit`

### ISSUE THREAD

**Issue by directhex** [emscripten-core/emsdk#1214]
I can understand most of the differences between the two, it's just basic directory layout stuff. What I'm confused by is the `libc++.a`, `libc++abi.a` and `libc++experimental.a` (along with all the libc++ headers). These are built for the host architecture (i.e. linux x64 or whatever), but building LLVM from source (`emsdk install llvm-git-main-64bit`) doesn't build/use/install libc++ for the host architecture.

```
$ readelf -h upstream/lib/libc++experimental.a | grep Machine
  Machine:                           Advanced Micro Devices X86-64
  Machine:                           Advanced Micro Devices X86-64
```

When would an end user need these in a binary release of emsdk? What am I missing? And how did they even get there, when `build_llvm` isn't passing `-DLLVM_ENABLE_RUNTIMES="libcxx"`?

**Comment by sbc100** [comment]
- `sdk-main-64bit` builds everything from source.  
- `latest` install binaries that are prebuilt by the CI systems.

To maximize compatibility the prebuilt binaries compile thier own version of libc++ (for use by the host llvm binaries) rather than depending on the host system to have a recent libc++ version installed.   I guess the "build from source" recipe in emsdk does not do this.

The prebuilt binaries are also build using a different script: https://chromium.googlesource.com/emscripten-releases/+/refs/heads/main/src/build.py.  So I guess its not surprising the some differences exist, although they are both simply building unmodified versions of upstream llvm tools.

**Comment by directhex** [comment]
> To maximize compatibility the prebuilt binaries compile thier own version of libc++ (for use by the host llvm binaries) rather than depending on the host system to have a recent libc++ version installed.

Okay, that would explain shipping a libc++.so for dynamic linking. When would an end user consume libc++.a from their binary release of emsdk, in order to statically link their own x64 code?

**Comment by sbc100** [comment]
I don't think there is any need for `upstream/lib/libc++.a` and those other libraries.  The binaries I guess are statically linked with those files, so that fact that we ship them at all is likely just a artifact of the way llvm's `make install` phase works.

@dschuff should to prune these from the archive to save a few bytes?

**Comment by directhex** [comment]
If Sent from my iPhoneOn May 9, 2023, at 6:18 PM, Sam Clegg ***@***.***> wrote:﻿
I don't think there is any need for upstream/lib/libc++.a and those other libraries.  The binaries I guess are statically linked with those files, so that fact that we ship them at all is likely just a artifact of the way llvm's make install phase works.
@dschuff should to prune these from the archive to save a few bytes?

**Comment by directhex** [comment]
If saving space is a goal, I've found a few more potential options.

- Dynamically link libbinaryen instead of statically linking it. This saves about 100 meg on disk on Mac/Linux (~10 meg per binary, with a single 20 meg libbinaryen.so shared between a dozen binaries). `rpath` is already set correctly to deal with this in the binaryen build system.
- Instead of shipping both clang.exe and clang++.exe, use `clang.exe --driver-mode=g++` (see https://github.com/llvm/llvm-project/blob/main/clang/lib/Driver/ToolChain.cpp#L196-L215 for known mappings). This saves about 70 meg on disk on Windows.

**Comment by sbc100** [comment]
Yes those saving seem worth looking into (maybe ~10%).  @dschuff WDYT?   I think we has some issues with dynamic linking in the past.. One downside would be that we wouldn't be able to take advantage of LTO in the binaries and I think we care more about performance that code size, at least for wasm-opt.

**Comment by directhex** [comment]
Could statically link that one and dynamically the rest. Still saves dozens of megs. I don't remember off the top of my head how amenable the binaryen cmake is to that

**Comment by dschuff** [comment]
I agree that there's probably no need to include the static host libc++.a and related files. I think we can just prune them.
For static vs dynamic linking and size-related issues, we actually have 2 different build modes on the emsdk CI: one that we use for releases, and one that we use for continuous builds. The tagged release builds (which the `latest` target should almost always point to) are optimized with LTO, and use static linking for both Binaryen and LLVM. That definitely comes at a size cost, but the performance benefits are large enough that I thought it was probably worth it (although I would love to hear from users if the size of the emsdk binary packages starts to become problematic). The continuous builds are optimized for faster build time of the toolchain itself, and smaller storage/upload size. (These can be installed with emsdk's `tot` target or by manually specifying an emscripten-releases hash. Most users won't do this unless they want to test bleeding-edge pre-release code or bisect a bug).

**Comment by sbc100** [comment]
I think this issue is still worth fixing.  We never got around to pruning libc++.a or duplicating the clang/clang++ binaries.
